### ISSUE TITLE
create/update/apply supports --exclude

### ISSUE THREAD

**Issue by heyjcollins** [vmware-tanzu/apps-cli-plugin#33]
### Description of problem
raw feedback below --- requires polishing.
File excludes; as more systems around us adopt the "as code" approaches, developers will increasingly have files in their projects that aren't anything do with actually running code; i.e. they dont end up in the running container this impacts the usability of --local-path as it sends the source more often than it needs to. This situation already exists actually as dev's often store project specific IDE configuration in their projects. Also, bare in mind the new backstage component.yaml/system.yaml. All nothing to do with running the app. but devs often change them and when they next run apply with -local-path=. publishes the source code image. But in reality the app source code hasn't actually changed.

As more systems around us adopt the "as code" approaches, application developers will increasingly have files in their projects that have nothing to do with actually running code (those files don't end up in the running container).

This impacts the usability of `--local-path` as it:
* consumes resources unnecessarily (uploads code that isn't required)
* triggers automatic deployments (when `--live-update` is enabled) whenever any of those unnecessary files change. 

### Proposed solution

```gherkin
Given I have an app project that includes extraneous `.foo` files 
When I create a workload with my project and enable `--live-update`
Then my `*.foo` files are not uploaded on initial creation and when I update any `.foo` file in my project, a live-update is not triggered.
```
The UX needs to be designed with @danfein (needs-design label applied).

### Describe alternatives you've considered

Decision on solution hasn't been made...
* use an --exclude flag on a per-command basis
* use a persistent .ignore file

#### Prior art for review:
- https://docs.github.com/en/get-started/getting-started-with-git/ignoring-files
- https://www.perforce.com/manuals/p4vs/Content/P4VS/managing.excluding.html
- https://tortoisesvn.net/docs/release/TortoiseSVN_en/tsvn-dug-ignore.html 
- https://marketplace.visualstudio.com/items?itemName=AdamWulkiewicz.ExcludeFromBuild
- https://www.jetbrains.com/help/idea/compiler-excludes.html

**Comment by heyjcollins** [comment]
comment originally submitted by @scothis:

We should explore this not as a cli flag, but instead as an ignore file the persistently defines file patterns to exclude.  We could introduce a `.tanzuignore` or fallback to a `.gitignore`.

**Comment by heyjcollins** [comment]
comment originally submitted by @paulcwarren:

Ping. Any update on this issue? This is going to hurt developer's quite badly IMO because a change to any file in their "project" causes a re-deployment and that is slow.

**Comment by danfein** [comment]
> comment originally submitted by @scothis:
> 
> We should explore this not as a cli flag, but instead as an ignore file the persistently defines file patterns to exclude. We could introduce a `.tanzuignore` or fallback to a `.gitignore`.

If folks are already ignoring infrastructure files it makes sense to me to pick up on the same mechanism as a starting point, rather than making them specify things multiple times.

**Comment by heyjcollins** [comment]
IDEs can/will make updates to IDE-specific files in a project (which have no impact to app features themselves) as a dev is working. Every time such an update occurs, this would trigger a live-update.

**Comment by scothis** [comment]
> IDEs can/will make updates to IDE-specific files in a project (which have no impact to app features themselves) as a dev is working. Every time such an update occurs, this would trigger a live-update.

The Tanzu cli has no direct role in watching the file system for changes or syncing them into the running container. This behavior is encapsulated within the live update agent (e.g. Tilt). It looks like [Tilt respects](https://docs.tilt.dev/file_changes.html#where-ignores-come-from) both `.dockerignore` and `.tiltignore`. Other agents will have different behavior.

cc @paulcwarren

**Comment by paulcwarren** [comment]
This isn't to do with live updating @scothis.  Live update works fine as it is configured to ignore everything except application files.    

This is about optimizing the developer's workflow more generally.  Tanzu cli doesn't know to ignore changes to things like the `.vscode/` or `.idea/` directories.  As a result it will upload source code far too often.

**Comment by heyjcollins** [comment]
I've had some clarifying conversations with @paulcwarren and will make some updates to the original issue based on our discussions.

**Comment by heyjcollins** [comment]
@scothis - I'm not necessarily advocating we do this but...
Yes, the apps plugin doesn't currently watch the file system. 
However, if watching the file system enables the plugin to increase the efficiency of the application developer, adding that capability is worth considering.

**Comment by gnivedita** [comment]
@heyjcollins I want to clarify this part of the statement above "triggers automatic, time consuming, deployments (when --live-update is enabled) whenever any of those unnecessary files change." to make sure I understand the issue.
I think there are two parts involved here, 1) uploading the changes in local file system to the image registry 2) then syncing the change (only .class files that changed) to running container. I think we are talking about optimizing the first part and not the second one. Because the second action ( the actual live update) does not get triggered unless there are some newer version of .class files, is this correct understanding ? cc @paulcwarren

**Comment by heyjcollins** [comment]
Yes @gnivedita - this issue is about optimizing the upload (meaning - make it possible to only invoke the supply chain if/when there's been a code change). The actual live-update is already optimized.

**Comment by heyjcollins** [comment]
@danfein - we've made a decision to go with (new)`.tanzuignore`.
There's a couple UX items we need to think though and provide.

1. how can we maximize the discoverability of this feature within the CLI itself?
1. what feedback can/should we provide in command output when files are being ignored?
1. how/where should we document the feature?

**Comment by heyjcollins** [comment]
Decisions on scope:
support relative paths to individual files
support relative path to directories

**Comment by heyjcollins** [comment]
## AC
**validated proper message displayed when .tanzuignore exists**
```
%tanzu apps workload apply app-no-ignore --local-path .  --type web -y
The files and/or directories listed in the .tanzuignore file are being excluded from the uploaded source code.
Publishing source in "." to "gcr.io/pivotal-knative/jc-apps:latest"...
...
```

**validated proper message displayed when .tanzuignore DOESN'T exist**
```
%tanzu apps workload apply app-no-ignore --local-path .  --type web -y
Publishing source in "." to "gcr.io/pivotal-knative/jc-apps:latest"...
Published source
...
```

**validated files/dirs included in .tanzuignore are not uploaded**<br/>
created workload, with `.tanzuignore` in place as follows:
```
.git
.github
.gitignore
LICENSE
README.md
Archive.zip
jc.txt
accelerator.yaml
Tiltfile
## directories
catalog
foo
```
then updated `jc.txt` and added new files to `foo` directory

then applied "changes" -- expect it to skip the update to the workload because only excluded files/directories have changed
```
%tanzu apps workload apply app-no-ignore --local-path .  --type web -y
The files and/or directories listed in the .tanzuignore file are being excluded from the uploaded source code.
Publishing source in "." to "gcr.io/pivotal-knative/jc-apps:latest"...
Published source
Workload is unchanged, skipping update
...
```

**validated that the .tanzuignore is "ignored" when not in root of project directory**
Moved .tanzuignore into a directory
executed the apply
confirmed the files that were previously being ignored, are now included which triggers and update through the supply chain
